src/compiler/program.ts

```diff
@@ -36,57 +36,369 @@ namespace ts {
+        }
+
+        if (compilerOptions.rootDirs) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.rootDirs_option_is_specified_searching_for_a_longest_matching_prefix);
+            }
+
+            let matchedPrefix: string;
+            for (const rootDir of compilerOptions.rootDirs) {
+                let normalizedRoot = getNormalizedAbsolutePath(rootDir, baseUrl);
+                if (!endsWith(normalizedRoot, directorySeparator)) {
+                    normalizedRoot += directorySeparator;
+                }
+                if (startsWith(candidate, normalizedRoot) && (matchedPrefix === undefined || matchedPrefix.length < normalizedRoot.length)) {
+                    matchedPrefix = normalizedRoot;
+                }
+            }
+            if (matchedPrefix) {
+                const suffix = candidate.substr(matchedPrefix.length);
+                if (traceEnabled) {
+                    trace(host, Diagnostics.Found_longest_matching_rootDir_0_converted_relative_path_1_to_non_relative_path_2, matchedPrefix, moduleName, suffix);
+                }
+
+                return baseUrlResolveNonRelativeModuleName(suffix, baseUrl, supportedExtensions, compilerOptions, host, traceEnabled);
+            }
+            return { resolvedModule: undefined, failedLookupLocations };
+        }
+        else {
+            if (traceEnabled) {
+                trace(host, Diagnostics.rootDirs_option_is_not_specified_using_0_as_candidate_location, candidate);
+            }
+
+            const resolvedFileName = loadModuleFromFile(supportedExtensions, candidate, failedLookupLocations, host, traceEnabled);
+            return {
+                resolvedModule: resolvedFileName ? { resolvedFileName } : undefined,
+                failedLookupLocations
+            };
+        }
+    }
+
+    function baseUrlResolveNonRelativeModuleName(moduleName: string, baseUrl: string, supportedExtensions: string[], compilerOptions: CompilerOptions, host: ModuleResolutionHost, traceEnabled: boolean): ResolvedModuleWithFailedLookupLocations {
+        let longestMatchPrefixLength = -1;
+        let matchedPattern: string;
+        let matchedStar: string;
+
+        const failedLookupLocations: string[] = [];
+        if (compilerOptions.paths) {
+            if (traceEnabled) {
+                trace(host, Diagnostics.paths_option_is_specified_looking_for_a_pattern_to_match_module_name_0, moduleName);
+            }
+
+            for (const key in compilerOptions.paths) {
+                const pattern: string = key;
+                const indexOfStar = pattern.indexOf("*");
+                if (indexOfStar !== -1) {
+                    const prefix = pattern.substr(0, indexOfStar);
+                    const suffix = pattern.substr(indexOfStar + 1);
+                    if (moduleName.length >= prefix.length + suffix.length &&

```

Comment:
Just to clarify, the `moduleName.length >= prefix.length + suffix.length` is more for optimization? because I think below caz should catch the case when `moduleName.length < prefix.length + suffix.length`? as `startsWith(moduleName, prefix)` will be false?